### ISSUE TITLE
fix(cli): improve messaging on renku installation update

### ISSUE THREAD

**Issue by jsam** [SwissDataScienceCenter/renku-python#1083]
When, local installation of renku is outdated, user is greeted with the following message:

```
You are using renku version 0.8.2, however version 0.9.1 is available.
You should consider upgrading ...
```

The suggestion we received from users is that we add to this message a command which they could use to do this. We should default to `pip` usage and recommend it as a default package manager for updating.

**Comment by rokroskar** [comment]
`pipx` not `pip` :)

**Comment by emmjab** [comment]
because it's kind of an important detail -- the experience is different if in notebooks or local;

in notebooks:
- image provides `pipx`
- if you try to install new version with `pip` instead and you don't update paths, you'll still be using old `renku` from `pipx`
- if you open a new notebook, you'll still be using old `renku`
- solution should involve updating the `FROM` statement in the docker file to use a later `renku`
- solution? can spit out the `FROM` line, path to the dockerfile, and link to the docs to help people update this
- maybe something more automated later, but still an issue if people are using their own base images

local:
- if your original installation was in `pipx` and you try to install new version with pip instead and you don't update paths, you'll still be using old version
- `pip` vs `pipx` caused me & others a lot of headaches, we should get our story straight and update the readme & docs

**Comment by rokroskar** [comment]
I would discourage anyone from installing it with `pip` if all they want is to use the CLI. If they want to use it as an sdk then it can be installed with `pip` but probably in a virtual environment.

**Comment by jsam** [comment]
@rokroskar interesting, could you explain your reasoning about this a bit more? 

I see it completely opposite and we should discourage people from using pipx since they are not sponsored by PSA or PSF, it's a relatively small project, most of the community is not using it and on top of that, it's notorious for causing installation trouble. I see no good reason for keeping this around as part of our official support.

**Comment by rokroskar** [comment]
yes - renku has a lot of dependencies, many of them are pinned to a specific version. This tends to cause unresolvable conflicts for users very quickly when they install it into their generic python environment. Our default installation channel was pip and at our first hackathon people's python environments were blowing up all over the place. For command-line tools, pip is a very bad choice imo, and your assumption about users who want to isolate tools in their own environments knowing what they are doing is not valid in my experience. 

Our users typically have no idea what a python environment is, where python is installed, or what isolation even means. `pipx` is simply a wrapper for `venv` and `pip`.

**Comment by jsam** [comment]
I understand, and I agree that on paper this idea is better - it's just that experience so far showed the opposite case. I've check out other CLI tools (gcloud, yt-download, ...)  based on Python and none of them is recommending pipx.

Considering that so far I had more than 3 cases where I had to troubleshoot user's issues which were all related to pipx and none of them were renku related - it seems that it just adds technical debt and not really viable long term solution. To me, it seems like a good short-sighted solution that works well for hackathons, but is potentially a quite bad solution for LTS (due to amount of problems we have already encountered with it just being in our readme). I'm happy to be proven otherwise. 

Until pipx becomes officially supported distribution method and general community accepted solution, my vote goes to pip.

**Comment by rokroskar** [comment]
What were you troubleshooting that was pipx-related? What ended up being the problem?

What do you mean by "since I see no problems with having renku versions per project - you don't have to pollute globally if you don't want to"?

`pipx` uses `pip` to install packages - so whatever dependency resolution issues you were having are related to `pip` and not `pipx`. All `pipx` does is create a virtual environment, use`pip` to install the package there and then make a symbolic link to the executable script from `~/.local/bin`.

**Comment by jsam** [comment]
The usage pattern I've noticed is that users install renku globally and then if they have renku project sometimes they install it in a project environment which depends on renku. This creates a usage problem since they get confused about which renku are they using in the end. In this case, if you upgrade through pipx, even if you upgrade successfully you will end up upgrading the wrong renku installation which ends up in more confusion.

I understand that internally it relies on pip (even though it does not ship with it), but upgrading packages which in new version adds more package ended up in incorrect environment set up (and this happened on multiple times on both linux and osx). 

I did not inspect it in detail so I cannot say if the problem was exactly due to transient dependencies or simply a failure in adding additional deps.

Either way, if we decide to support this, we should look into how to resolve all of those issues we are buying into. Also editing readme files with explaining most common pitfalls would be a good idea.

**Comment by rokroskar** [comment]
It would be great if you can reproduce the problem and document it - I've been using `pipx` happily for 2 years without problems. They did change their CLI a bit recently to be more in line with `pip` which was annoying but not a deal-breaker for me.

**Comment by jsam** [comment]
Just to be clear on what I'm talking about - our default distribution method should be pip and we should recommend pip as our default. 

This does not mean, that I'm against on having a section which describes global installation of renku through pipx - on contrary I think we should have it, it's just I don't think it should be the first line for all the python newbies out there. As a starting out guide and a default management of renku I would recommend users that they install renku as a dependency in their Python projects. 


For non-python base projects and multi-language support (R, Ruby, JS) we will most likely need to build global installation support in which we will probably have to either depend on docker or system init which is being used - but that's a different discussion for another time.

**Comment by jsam** [comment]
Sure. I think for starters, we should improve our docs on that part since most of the usage confusion comes from the case I mentioned. For example, pipx section could be moved into a section called "global installation" which explains global vs local installations and then adding a section which explains what and how to use renku if you install it as a dependency. I think the latter is completely missing currently.

**Comment by rokroskar** [comment]
The problem you describe sounds like a generic virtual environment issue. I'm also surprised this is common since most people don't use renku as a library but strictly as a cli. I wouldn't confuse the docs with "global" etc sections - there should be a super short install section for the most common use case (renku as cli) and maybe another section that mentions specifically usage as a library. This is basically what I wrote in #1098.

**Comment by rokroskar** [comment]
FWIW `pipx` has moved much closer to `pip` in the months since this issue was first opened. They basically follow the `pip` syntax now so showing this in a brief message as suggested at the top of this issue seems more plausible now.
